sdk/metric/controller/basic/controller.go

```diff
@@ -0,0 +1,302 @@
+			return
+		case <-c.ticker.C():
+			if err := c.collect(ctx); err != nil {
+				otel.Handle(err)
+			}
+		}
+	}
+}
+
+// collect computes a checkpoint and optionally exports it.
+func (c *Controller) collect(ctx context.Context) error {
+	if err := c.checkpoint(ctx, func() bool {
+		return true
+	}); err != nil {
+		return err
+	}
+	if c.exporter == nil {
+		return nil
+	}
+	// This export has to complete before starting another
+	// collection since it will hold a read lock and
+	// checkpoint() must re-acquire a write lock.
+	if err := c.export(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkpoint calls the Accumulator and Checkpointer interfaces to
+// compute the CheckpointSet.  This applies the configured collection
+// timeout.  Note that this does not try to cancel a Collect or Export
+// when Stop() is called.
+func (c *Controller) checkpoint(ctx context.Context, cond func() bool) error {
+	ckpt := c.checkpointer.CheckpointSet()
+	ckpt.Lock()
+	defer ckpt.Unlock()
+
+	if !cond() {
+		return nil
+	}
+	c.checkpointer.StartCollection()
+
+	if c.timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	_ = c.accumulator.Collect(ctx)
+
+	var err error
+	select {
+	case <-ctx.Done():
+		err = ctx.Err()
+	default:
+		// The context wasn't done, ok.
+	}
+
+	// Finish the checkpoint whether the accumulator timed out or not.
+	if cerr := c.checkpointer.FinishCollection(); cerr != nil {
+		if err == nil {
+			err = cerr
+		} else {
+			err = fmt.Errorf("%s: %w", cerr.Error(), err)
+		}
+	}
+
+	return err
+}
+
+// export calls the exporter with a read lock on the CheckpointSet,
+// applying the configured export timeout.
+func (c *Controller) export() error {
+	ckpt := c.checkpointer.CheckpointSet()
+	ckpt.RLock()
+	defer ckpt.RUnlock()
+
+	ctx := context.Background()
+	if c.exportTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, c.exportTimeout)
+		defer cancel()
+	}
+
+	return c.exporter.Export(ctx, ckpt)
+}
+
+// Foreach gives the caller read-locked access to the current
+// export.CheckpointSet.
+func (c *Controller) ForEach(ks export.ExportKindSelector, f func(export.Record) error) error {
+	ckpt := c.checkpointer.CheckpointSet()
+	ckpt.RLock()
+	defer ckpt.RUnlock()
+
+	return ckpt.ForEach(ks, f)
+}
+
+// IsRunning returns true if the controller was started via Start(),
+// indicating that the current export.CheckpointSet is being kept
+// up-to-date.
+func (c *Controller) IsRunning() bool {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	return c.ticker != nil
+}
+
+// Collect requests a collection.  The collection will be skipped if
+// the last collection is aged less than the configured collection
+// period.
+func (c *Controller) Collect(ctx context.Context) error {
+	if c.IsRunning() {
+		// When there's a non-nil ticker, there's a goroutine
+		// computing checkpoints with the collection period.
+		return nil
```

Comment:
This seems like this would only result from a misuse of the controller interface (calling `Collect` after `Start`)).  Should we return an error here?  Otherwise, as a user I would expect Collect to trigger an additional collection in addition to the ticker-based one.

Reply:
It returns an error now. Thanks!